### PR TITLE
Add optional wrappers (V2) to AOT headers to take Halide::Runtime::Bu…

### PR DESCRIPTION
**Pull request by steven-johnson**
…ffers as arguments

This is a reworked variant of #3206 that dispenses with the Google-style convention and requires all Buffer arguments to be mutable, whether by ref or ptr. Note also that (unlike the previous patch) it disables the implicit `operator halide_buffer_t*` provided by the `Buffer` class, in an effort to ensure that it's harder to pass a wrong-static-type Buffer to a given function (and, of course, with the idea that implicit conversions are always preferable to avoid when not absolutely necessary).

Not sure if this is ready or desirable to land, but putting it out there for discussion purposes.

### CODE REVIEW

**Comment by steven-johnson** [comment]
Brought up to date and tweaked a little, but this is basically the same as it was before. I think it's worth considering. PTAL.

**Comment by steven-johnson** [comment]
Resurrected and simplified to try to address https://github.com/halide/Halide/issues/3955. Thoughts?

File: src/CodeGen_C.cpp [hunk 1]
```diff
@@ -1787,6 +1829,44 @@ void CodeGen_C::compile(const LoweredFunc &f) {
 
         // And also the metadata.
         stream << "const struct halide_filter_metadata_t *" << simple_name << "_metadata() HALIDE_FUNCTION_ATTRS;\n";
+
+        // Emit overload(s) that accept Halide::Runtime::Buffers instead of halide_buffer_t
+        // (but only if the function is being emitted in NameMangling::CPlusPlus mode)
+        if (name_mangling == NameMangling::CPlusPlus) {
```
Comment (abadams) [inline src/CodeGen_C.cpp]: You can overload a C-linkage function with a C++ linkage function right? I wonder if we really want this limit.

**Comment by abadams** [comment]
I think this is a good idea. I just question whether or not we want to tie it to C++ name mangling.

**Comment by steven-johnson** [comment]
> whether or not we want to tie it to C++ name mangling

I suspect the idea was to maintain plain-C compatibility, in case anyone needs to use the header in a non-C++ world, where the overloads aren't legal. (May be moot in this case since it's opt-in via #define).

That said, I'm unsatisfied with an opt-in approach; IMHO it would be better to have generation work this way "out of the box".

**Comment by abadams** [comment]
I meant that the halide_buffer_t entrypoint could be extern "C", and we could still have C++ overloads for it. I think that works.

**Comment by steven-johnson** [comment]
Ah, yes, of course, that would work. That still leaves us with the opt-in requirement, which isn't horrible but definitely suboptimal IMHO.

**Comment by abadams** [comment]
How about we make it opt-in for now, try to port a bunch of production code over to it, and if it goes well remove the opt-in ifdef and mark the halide_buffer_t implicit cast as deprecated, and then eventually delete it.

**Comment by steven-johnson** [comment]
> How about we make it opt-in for now

SGTM -- let me do a test in our codebase before we land this.

**Comment by steven-johnson** [comment]
(BTW, the reason for the opt-in is to remove the need to include HalideBuffer.h before these files; that would still be an issue)

**Comment by abadams** [comment]
A way to dodge the include would be to template the wrapper on the entire buffer type instead of just the element type, and just require that it's something with an 'as' method and a 'raw_buffer' method. This would additionally enable type-checking for people who have an existing buffer class they want to be able to pass into Halide. We could even roll the as/raw_buffer combo into a `check_type_and_return_raw_buffer` method to simplify things for those users.

**Comment by dsharletg** [comment]
> We could even roll the as/raw_buffer combo into a check_type_and_return_raw_buffer method to simplify things for those users.

If we made this a free function instead of a member function, it could be used to enable passing custom buffer types that the user doesn't control too: an overload could be provided by user code for any type.

**Comment by dsharletg** [comment]
I think that would make it easier to solve the forward declaration problem too.

**Comment by steven-johnson** [comment]
> If we made this a free function 

I like it. Stand by.

**Comment by abadams** [comment]
It would be even better if it somehow had the choice between returning a pointer to an existing halide_buffer_t or populating a temporary one. I guess it could return a wrapper class by value with an implicit cast method (i.e. the user's overload could just do what H::R::Buffer currently does), so we should avoid constraining the return type of the free function. Is there a cleaner solution?

**Comment by steven-johnson** [comment]
> populating a temporary one

Feature creep -- I think that requiring any bufferlike class to provide a longterm halide_buffer_t isn't unreasonable here.

**Comment by steven-johnson** [comment]
(that said, I think I can it work.)

**Comment by abadams** [comment]
Acknowledged, but if it's simple to not constrain the return type of the free function we will increase flexibility for people down the line.

**Comment by steven-johnson** [comment]
OK, I think this accomplishes what has been requested -- it *should* boil down to negligible code for the common case, but I still need to verify. PTAL.

**Comment by steven-johnson** [comment]
I verified that on x86-64-osx, clang with -O2 correctly boils down the `halide_buffer_accessor_t` nonsense into a single `leaq` if the Buffer type is statically known. (If it is Buffer<void> then a runtime type check is also done.)

**Comment by steven-johnson** [comment]
Meh, there's a fly in this ointment: the way I've done this requires that the arguments be lvalues, which may not be the case (e.g. if they are temporary wrappers around something else). Perhaps just adding an overload for pass-by-value would work.

**Comment by steven-johnson** [comment]
This is ~never going to land in this form. Closing.

**Pull request closed without merging**